src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+-- and receipt associated to a block in the blockchain.
+--
+-- This module should be imported qualified.
+module Oscoin.Storage.Ledger
+    ( Ledger
+    , hoist
+    , mkLedger
+    , newMemoryIO
+    , newMemoryFromGenesisIO
+    , newFromBlockStore
+
+    , blockStoreReader
+    , blockStore
+    , getBlocksByDepth
+    , getLatestBlockWithState
+    , getTip
+    , getState
+    , getReceipt
+
+    , buildNextBlock
+    , addBlock
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Blockchain.Eval
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Storage.Block.Abstract (BlockStore)
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Memory
+import           Oscoin.Storage.HashStore
+import qualified Oscoin.Storage.Receipt as ReceiptStore
+import qualified Oscoin.Time.Chrono as Chrono
+
+import           Codec.Serialise (Serialise)
+import           Control.Monad.Trans.Maybe
+import qualified Crypto.Data.Auth.Tree.Class as AuthTree
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | Errors that maybe raised from the ledgers methods
+data LedgerError c
+    = StateNotInCache (BlockHash c)
+    -- ^ State for the given block is not materialized in the state
+    -- cache.
+
+deriving instance (Crypto.HasHashing c) => Eq (LedgerError c)
+deriving instance (Show (Crypto.Hash c)) => Show (LedgerError c)
+
+data Ledger crypto seal tx output state (m :: * -> *) = Ledger
+    { blockStore   :: BlockStore crypto tx seal m
+    , stateStore   :: HashStore crypto state m
+    , receiptStore :: ReceiptStore.ReceiptStore crypto tx output m
+    , ledgerEval   :: Evaluator state tx output
+    }
+
+hoist
+    :: (forall a. m a -> n a)
+    -> Ledger crypto seal tx output state m
+    -> Ledger crypto seal tx output state n
+hoist f l =
+    Ledger
+        { blockStore = BlockStore.hoistBlockStore f $ blockStore l
+        , stateStore = hoistHashStore f $ stateStore l
+        , receiptStore = ReceiptStore.hoistReceiptStore f $ receiptStore l
+        , ledgerEval = ledgerEval l
+        }
+
+
+-- | Create a ledger using in-memory storage for blocks, states, and
+-- receipts.
+--
+-- Uses 'newFromBlockStore' with 'newBlockStoreIO' initialized with the
+-- given blocks.
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newMemoryIO
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> Chrono.NewestFirst NonEmpty (Block crypto tx (Sealed crypto seal))
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryIO ledgerEval blks initialState = do
+    blockStore <- newBlockStoreIO blks
+    newFromBlockStore ledgerEval blockStore initialState
+
+
+-- | Same as 'newMemoryIO' with just @genesisBlk@ in the blockchain.
```

Comment:
The function parameter names won't be visible in the haddock, so maybe you could say "the given block"